lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -2455,6 +2466,44 @@ ParseResult FIRStmtParser::parseListExp(Value &result) {
   return success();
 }
 
+/// list-concat-exp ::= 'list_concat' '(' exp* ')'
+ParseResult FIRStmtParser::parseListConcatExp(Value &result) {
+  consumeToken(FIRToken::lp_list_concat);
+
+  auto loc = getToken().getLoc();
+  ListType type;
+  SmallVector<Value, 3> operands;
+  if (parseListUntil(FIRToken::r_paren, [&]() -> ParseResult {
+        Value operand;
+        locationProcessor.setLoc(loc);
+        if (parseExp(operand, "expected expression in List concat expression"))
+          return failure();
+
+        if (!type_isa<ListType>(operand.getType()))
+          return emitError(loc, "unexpected expression of type ")
+                 << operand.getType() << " in List concat expression";
+
+        if (!type)
+          type = type_cast<ListType>(operand.getType());
+
+        if (operand.getType() != type)
+          return emitError(loc, "unexpected expression of type ")
+                 << operand.getType() << " in List concat expression of type "
+                 << type;
+
+        operands.push_back(operand);
+        return success();
+      }))
+    return failure();
+
+  if (operands.empty())
+    return emitError(loc, "need at least one List to concatenate");
```

Comment:
This is legal syntactically according to the spec, but for this specific operation, the IR requires at least one operand. It makes the parser simpler to implement by requiring at least one operand and inferring the type from the types of the operands, but if it would be preferable to not require this during parsing (and instead let the op's verifier enforce it), I can change that.